File: TCProject/src/CityMap.java
import java.util.regex.*;
import static java.lang.Math.*;
import static java.util.Arrays.*;
import static java.lang.Integer.*;
import static java.lang.Double.*;
import static java.util.Collections.*;
import java.util.*;

public class CityMap {
	public String getLegend(String[] cityMap, int[] POIs) {
		String res = "";
		HashMap<Character, Integer> map = new HashMap<Character, Integer>();
		for (int j = 0; j < cityMap.length; j++) {
			for (int j2 = 0; j2 < cityMap[j].length(); j2++) {
				char c = cityMap[j].charAt(j2);
				if (map.containsKey(c)) {
					map.put(c, map.get(c) + 1);
				} else {
					map.put(c, 1);
				}
			}
		}
		pr(map);
		for (int i = 0; i < POIs.length; i++) {
			for (Character s : map.keySet()) {
				if(map.get(s)==POIs[i]){
					res+=s;
					break;
				}
			}
		}
		return res;
	}

	// BEGIN CUT HERE

	public static void main(String[] args) {
		try {
			eq(0,
					(new CityMap()).getLegend(new String[] { "M....M",
							"...R.M", "R..R.R" }, new int[] { 3, 4 }), "MR");
			eq(1, (new CityMap()).getLegend(new String[] { "XXXXXXXZXYYY" },
					new int[] { 1, 8, 3 }), "ZXY");
			eq(2, (new CityMap()).getLegend(new String[] { "...........",
					"...........", "...........", "..........T" },
					new int[] { 1 }), "T");
			eq(3, (new CityMap()).getLegend(new String[] {
					"AIAAARRI.......GOAI.", ".O..AIIGI.OAAAGI.A.I",
					".A.IAAAARI..AI.AAGR.", "....IAI..AOIGA.GAIA.",
					"I.AIIIAG...GAR.IIAGA", "IA.AOA....I....I.GAA",
					"IOIGRAAAO.AI.AA.RAAA", "AI.AAA.AIR.AGRIAAG..",
					"AAAAIAAAI...AAG.RGRA", ".J.IA...G.A.AA.II.AA" },
					new int[] { 16, 7, 1, 35, 11, 66 }), "GOJIRA");
		} catch (Exception exx) {
			System.err.println(exx);
			exx.printStackTrace(System.err);
		}
	}

	private static void pr(Object... rs) {
		System.err.println(Arrays.deepToString(rs).replace("]", "]\n"));
	}

	private static void eq(int n, int a, int b) {
		if (a == b) {
			System.err.println("Case " + n + " passed.");
		} else {
			System.err.println("Case " + n + " failed: " + received + " " + a
					+ " " + expected + " " + b + ".");
		}
	}

	private static void eq(int n, double a, double b) {
		if (a == b) {
			System.err.println("Case " + n + " passed.");
		} else {
			System.err.println("Case " + n + " failed: " + received + " '" + a
					+ "' " + expected + " '" + b + "'.");
		}
	}

	private static void eq(int n, char a, char b) {
		if (a == b) {
			System.err.println("Case " + n + " passed.");
		} else {
			System.err.println("Case " + n + " failed: " + received + " '" + a
					+ "' " + expected + " '" + b + "'.");
		}
	}

	private static void eq(int n, long a, long b) {
		if (a == b) {
			System.err.println("Case " + n + " passed.");
		} else {
			System.err.println("Case " + n + " failed: " + received + " '" + a
					+ "L' " + expected + " '" + b + "L'.");
		}
	}

	private static void eq(int n, boolean a, boolean b) {
		if (a == b) {
			System.err.println("Case " + n + " passed.");
		} else {
			System.err.println("Case " + n + " failed: " + received + " '" + a
					+ "' " + expected + " '" + b + "'.");
		}
	}

	private static void eq(int n, String a, String b) {
		if (a != null && a.equals(b)) {
			System.err.println("Case " + n + " passed.");
		} else {
			System.err.println("Case " + n + " failed: " + received + " \"" + a
					+ "\" " + expected + " \"" + b + "\".");
		}
	}

	private static void eq(int n, int[] a, int[] b) {
		if (a.length != b.length) {
			System.err.println("Case " + n + " failed: different lengths");
			print(received + "<" + a.length + "> ", a);
			print(expected + "<" + b.length + "> ", b);
			return;
		}
		for (int i = 0; i < a.length; i++) {
			if (a[i] != b[i]) {
				System.err
						.println("Case " + n + " failed. " + received + " and "
								+ expected + " array differ in position " + i);
				print(received + " ", a);
				print(expected + " ", b);
				return;
			}
		}
		System.err.println("Case " + n + " passed.");
	}

	private static void eq(int n, long[] a, long[] b) {
		if (a.length != b.length) {
			System.err.println("Case " + n + " failed: different lengths");
			print(received + "<" + a.length + "> ", a);
			print(expected + "<" + b.length + "> ", b);
			return;
		}
		for (int i = 0; i < a.length; i++) {
			if (a[i] != b[i]) {
				System.err
						.println("Case " + n + " failed. " + received + " and "
								+ expected + " array differ in position " + i);
				print(received, a);
				print(expected, b);
				return;
			}
		}
		System.err.println("Case " + n + " passed.");
	}

	private static void eq(int n, double[] a, double[] b) {
		if (a.length != b.length) {
			System.err.println("Case " + n + " failed: different lengths");
			print(received + "<" + a.length + "> ", a);
			print(expected + "<" + b.length + "> ", b);
			return;
		}
		for (int i = 0; i < a.length; i++) {
			if (a[i] != b[i]) {
				System.err
						.println("Case " + n + " failed. " + received + " and "
								+ expected + " array differ in position " + i);
				print(received, a);
				print(expected, b);
				return;
			}
		}
		System.err.println("Case " + n + " passed.");
	}

	private static void eq(int n, String[] a, String[] b) {
		if (a.length != b.length) {
			System.err.println("Case " + n + " failed: different lengths");
			print(received + "<" + a.length + "> ", (Object[]) a);
			print(expected + "<" + b.length + "> ", (Object[]) b);
			return;
		}
		for (int i = 0; i < a.length; i++) {
			if (!a[i].equals(b[i])) {
				System.err
						.println("Case " + n + " failed. " + received + " and "
								+ expected + " array differ in position " + i);
				print(received, (Object[]) a);
				print(expected, (Object[]) b);
				return;
			}
		}
		System.err.println("Case " + n + " passed.");
	}

	static String expected = "  expe";
	static String received = "  rc";
	// END CUT HERE
}

// BEGIN CUT HERE

// END CUT HERE

File: TCProject/src/cpp/CityMap.cpp
#include <algorithm>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <queue>
#include <set>
#include <map>
#include <cstdio>
#include <cstdlib>
#include <cctype>
#include <cmath>
using namespace std;
#define all(v) (v).begin(),(v).end()
#define rall(v) (v).rbegin(),(v).rend()
#define debug(a) cerr << #a << ": " << a << endl;
#define debugv(b) cerr << #b << ":\n"; for(int countvec = 0; countvec < b.size(); countvec++) {cerr << b[countvec] << "    ";} cerr << endl;
#define debuga(b) cerr << #b << ":\n"; for(int countvec = 0; countvec < sizeof(b)/sizeof(int); countvec++) {cerr << b[countvec] << "    ";} cerr << endl;
#define debugm(c) cerr << #c << ":\n"; for(int countmat = 0; countmat < c.size(); countmat++) {for(int countbb = 0; countbb < c[countmat].size(); countbb++) { cerr << c[countmat][countbb] << "    "; } cerr << endl; }
#define fill(a , value) memset(a , value , sizeof(a))
string tos(int a) {ostringstream os(""); os << a; return os.str();}


// BEGIN CUT HERE
vector<string> split( const string& s, const string& delim =" " ) {
    vector<string> res;
    string t;
    for ( int i = 0 ; i != s.size() ; i++ ) {
        if ( delim.find( s[i] ) != string::npos ) {
            if ( !t.empty() ) {
                res.push_back( t );
                t = "";
            }
        } else {
            t += s[i];
        }
    }
    if ( !t.empty() ) {
        res.push_back(t);
    }
    return res;
}

vector<int> splitInt( const string& s, const string& delim =" " ) {
    vector<string> tok = split( s, delim );
    vector<int> res;
    for ( int i = 0 ; i != tok.size(); i++ )
        res.push_back( atoi( tok[i].c_str() ) );
    return res;
}
// END CUT HERE

// BEGIN CUT HERE
#define ARRSIZE(x) (sizeof(x)/sizeof(x[0]))

template<typename T> void print( T a ) {
    cerr << a;
}
static void print( long long a ) {
    cerr << a << "L";
}
static void print( string a ) {
    cerr << '"' << a << '"';
}
template<typename T> void print( vector<T> a ) {
    cerr << "{";
    for ( int i = 0 ; i != a.size() ; i++ ) {
        if ( i != 0 ) cerr << ", ";
        print( a[i] );
    }
    cerr << "}" << endl;
}
template<typename T> void eq( int n, T have, T need ) {
    if ( have == need ) {
        cerr << "Case " << n << " passed." << endl;
    } else {
        cerr << "Case " << n << " failed: expected ";
        print( need );
        cerr << " received ";
        print( have );
        cerr << "." << endl;
    }
}
template<typename T> void eq( int n, vector<T> have, vector<T> need ) {
    if( have.size() != need.size() ) {
        cerr << "Case " << n << " failed: returned " << have.size() << " elements; expected " << need.size() << " elements.";
        print( have );
        print( need );
        return;
    }
    for( int i= 0; i < have.size(); i++ ) {
        if( have[i] != need[i] ) {
            cerr << "Case " << n << " failed. Expected and returned array differ in position " << i << ".";
            print( have );
            print( need );
            return;
        }
    }
    cerr << "Case " << n << " passed." << endl;
}
static void eq( int n, string have, string need ) {
    if ( have == need ) {
        cerr << "Case " << n << " passed." << endl;
    } else {
        cerr << "Case " << n << " failed: expected ";
        print( need );
        cerr << " received ";
        print( have );
        cerr << "." << endl;
    }
}
// END CUT HERE
class CityMap {
public:
    string getLegend(vector <string> cityMap, vector <int> POIs) {
        string res="";
        stringstream ss;
        int count = POIs.size();
        int map[26];
        fill(map,0);
        // debuga(map);
        for (int i = 0; i < cityMap.size(); ++i)
        {

            for (int j = 0; j < cityMap[i].length(); ++j)
            {
                if(cityMap[i][j]!='.'){
                    // debug((int)cityMap[i][j]-'A');
                    // debug(cityMap[i][j]);
                    map[cityMap[i][j]-'A']++;
                }
                    // map[]++;
            }
        }
        // debuga(map);
        for (int i = 0; i < count; ++i)
        {
            for (int j = 0; j < 26; ++j)
            {
                if(map[j]==POIs[i]){
                    // ss<<;
                    res+=(j+'A');
                    // ss>>res;
                }
            }
        }
        return res;
    }

};
// BEGIN CUT HERE
int main( int argc, char* argv[] ) {
    {
        string cityMapARRAY[] = {"M....M",
            "...R.M",
            "R..R.R"};
        vector <string> cityMap( cityMapARRAY, cityMapARRAY+ARRSIZE(cityMapARRAY) );
        int POIsARRAY[] = {3, 4};
        vector <int> POIs( POIsARRAY, POIsARRAY+ARRSIZE(POIsARRAY) );
        CityMap theObject;
        eq(0, theObject.getLegend(cityMap, POIs),"MR");
    }
    {
        string cityMapARRAY[] = {"XXXXXXXZXYYY"};
        vector <string> cityMap( cityMapARRAY, cityMapARRAY+ARRSIZE(cityMapARRAY) );
        int POIsARRAY[] = {1, 8, 3};
        vector <int> POIs( POIsARRAY, POIsARRAY+ARRSIZE(POIsARRAY) );
        CityMap theObject;
        eq(1, theObject.getLegend(cityMap, POIs),"ZXY");
    }
    {
        string cityMapARRAY[] = {"...........",
            "...........",
            "...........",
            "..........T"};
        vector <string> cityMap( cityMapARRAY, cityMapARRAY+ARRSIZE(cityMapARRAY) );
        int POIsARRAY[] = {1};
        vector <int> POIs( POIsARRAY, POIsARRAY+ARRSIZE(POIsARRAY) );
        CityMap theObject;
        eq(2, theObject.getLegend(cityMap, POIs),"T");
    }
    {
        string cityMapARRAY[] = {"AIAAARRI.......GOAI.",
            ".O..AIIGI.OAAAGI.A.I",
            ".A.IAAAARI..AI.AAGR.",
            "....IAI..AOIGA.GAIA.",
            "I.AIIIAG...GAR.IIAGA",
            "IA.AOA....I....I.GAA",
            "IOIGRAAAO.AI.AA.RAAA",
            "AI.AAA.AIR.AGRIAAG..",
            "AAAAIAAAI...AAG.RGRA",
            ".J.IA...G.A.AA.II.AA"}
           ;
        vector <string> cityMap( cityMapARRAY, cityMapARRAY+ARRSIZE(cityMapARRAY) );
        int POIsARRAY[] = {16,7,1,35,11,66}
           ;
        vector <int> POIs( POIsARRAY, POIsARRAY+ARRSIZE(POIsARRAY) );
        CityMap theObject;
        eq(3, theObject.getLegend(cityMap, POIs),"GOJIRA");
    }
}
// END CUT HERE
